src/expression.js

```diff
@@ -206,7 +206,7 @@ pp.parseExprSubscripts = function(refShorthandDefaultPos) {
   let startPos = this.start, startLoc = this.startLoc
   let expr = this.parseExprAtom(refShorthandDefaultPos)
   if (refShorthandDefaultPos && refShorthandDefaultPos.start) return expr
-  return this.parseSubscripts(expr, startPos, startLoc)
+  return this.parseSubscripts(expr, startPos, startLoc, expr.type === "ArrowFunctionExpression" && this.parenthesesCount === 0)
```

Comment:
Still thinking on whether we should pass this as `noCalls` or rather just not go into parsing subscripts at all. I suppose `foo => bar.x`, `` foo => bar`x` `` and `foo => bar[x]` are also already covered by body parsing so we can safely return at the point after unparenthesized arrow function, right?